src/box/lua/trigger.c

```diff
@@ -0,0 +1,242 @@
+ * arguments. Execution is stopped by a first exception.
+ * First argument must be a string - all the other arguments will be passed
+ * to the triggers without any processing or copying.
+ * Returns no values on success. If one of the triggers threw an error, it is
+ * raised again.
+ */
+static int
+luaT_trigger_call(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	struct event *event = event_get(event_name, false);
+	if (event == NULL)
+		return 0;
+	int top = lua_gettop(L);
+	int narg = top - 1;
+	struct event_trigger_iterator it;
+	event_trigger_iterator_create(&it, event);
+	struct func_adapter *trg = NULL;
+	const char *name = NULL;
+	int rc = 0;
+	while (rc == 0 && event_trigger_iterator_next(&it, &trg, &name)) {
+		lua_settop(L, top);
+		func_adapter_lua_get_func(trg, L);
+		for (int i = top - narg + 1; i <= top; ++i)
+			lua_pushvalue(L, i);
+		rc = luaT_call(L, narg, LUA_MULTRET);
+	}
+	event_trigger_iterator_destroy(&it);
+	if (rc != 0)
+		return luaT_error(L);
+	return 0;
+}
+
+/**
+ * Sets an array of arrays [trigger_name, trigger_handler] by event->name key
+ * in a pre-created table. Never sets an empty array.
+ */
+static bool
+trigger_info_push_event(struct event *event, void *arg)
+{
+	lua_State *L = (lua_State *)arg;
+	struct event_trigger_iterator it;
+	event_trigger_iterator_create(&it, event);
+	struct func_adapter *trg = NULL;
+	const char *name = NULL;
+	int idx = 0;
+	lua_createtable(L, 0, 0);
+	while (event_trigger_iterator_next(&it, &trg, &name)) {
+		idx++;
+		lua_createtable(L, 2, 0);
+		lua_pushstring(L, name);
+		lua_rawseti(L, -2, 1);
+		func_adapter_lua_get_func(trg, L);
+		lua_rawseti(L, -2, 2);
+		lua_rawseti(L, -2, idx);
+	}
+	lua_setfield(L, -2, event->name);
+	return true;
+}
+
+/**
+ * Pushes a key-value table, where the key is the event name and value is an
+ * array of triggers, represented by two-element [trigger_name, trigger_handler]
+ * arrays, registered on this event, in the order in which they will be called.
+ * If an event name is passed, a table contains only one key which is passed
+ * argument, if there is an event with such a name, or returned table is empty,
+ * if the event does not exist.
+ */
+static int
+luaT_trigger_info(struct lua_State *L)
+{
+	if (lua_gettop(L) == 0) {
+		lua_createtable(L, 0, 0);
+		bool ok = event_foreach(trigger_info_push_event, L);
+		assert(ok);
+		(void)ok;
+	} else {
+		const char *event_name = luaL_checkstring(L, 1);
+		struct event *event = event_get(event_name, false);
+		if (event == NULL || !event_has_triggers(event)) {
+			lua_createtable(L, 0, 0);
+			return 1;
+		}
+		lua_createtable(L, 0, 1);
+		trigger_info_push_event(event, L);
+	}
+	return 1;
+}
+
+const char *event_trigger_iterator_typename = "trigger.event_trigger_iterator";
+
+/**
+ * Lua-iterator over triggers of the event.
+ */
+struct lua_event_trigger_iterator {
+	/** An iterator over triggers of the event. */
+	struct event_trigger_iterator it;
+};
```

Comment:
Why do you need this structure? Can't you use `event_trigger_iterator` directly?

Reply:
Removed this structure.